DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionDataDelegateSwizzler.swift

```diff
@@ -0,0 +1,100 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+/// Swizzles `URLSessionDataDelegate` callbacks.
+internal class URLSessionDataDelegateSwizzler {
+    static var didReceiveMap: [String: DidReceive?] = [:]
+    static var lock = NSRecursiveLock()
+
+    static var isBinded: Bool {
+        lock.lock()
+        defer { lock.unlock() }
+        return didReceiveMap.isEmpty == false
+    }
+
+    static func bindIfNeeded(delegate: URLSessionDataDelegate, interceptDidReceive: @escaping (URLSession, URLSessionDataTask, Data) -> Void) throws {
+        lock.lock()
+        defer { lock.unlock() }
+
+        guard isBinded == false else {
+            return
+        }
+
+        try bind(delegate: delegate, interceptDidReceive: interceptDidReceive)
+    }
+
+    static func bind(delegate: URLSessionDataDelegate, interceptDidReceive: @escaping (URLSession, URLSessionDataTask, Data) -> Void) throws {
+        lock.lock()
+        defer { lock.unlock() }
+
+        let didReceive = try DidReceive.build(klass: type(of: delegate))
```

Comment:
Let's discuss on this, during the RFC review we decided to have the instance because of strongly typeness we get but I'm happy to hear what do you think or you know some alternative where customers can have benefit of both worlds.